python/ray/data/datasource/lancedb_datasource.py

```diff
@@ -0,0 +1,87 @@
+"""
+Module to read an LanceDB table into a Ray Dataset, by using the Ray Datasource API.
+"""
+
+import logging
+from typing import List, Optional, Union
+
+import lance
+import pyarrow as pa
+from lance import LanceFragment
+
+from ray.data import ReadTask
+from ray.data.block import Block, BlockMetadata
+from ray.data.datasource import Datasource
+from ray.util.annotations import DeveloperAPI
+
+logger = logging.getLogger(__name__)
+
+
+@DeveloperAPI
+class LanceDBDatasource(Datasource):
+    """LanceDB Datasource
+    Read a LanceDB table as a Ray Dataset
+
+    Parameters
+    ----------
+    uri : str
+        The base URI of the LanceDB dataset.
+    columns: list
+        A list of columns to return from the dataset.
+    filter: str
+        A standard SQL expressions as predicates for dataset filtering.
+    """
+
+    def __init__(
+        self,
+        uri: str,
+        columns: Union[list, None] = None,
+        filter: Union[str, None] = None,
+    ):
+        self.uri = uri
+        self.columns = columns
+        self.filter = filter
+
+        self.lance_ds = lance.dataset(uri)
+        self.fragments = self.lance_ds.get_fragments()
+
+    def get_read_tasks(self, parallelism: int) -> List[ReadTask]:
+        # To begin with, read one Fragment at a time
+        # Each Ray Data Block contains a Pandas RecordBatch
+        def _read_single_fragment(fragment: LanceFragment) -> Block:
+            # Fetch batches from the fragment
+            batches = fragment.to_batches(columns=self.columns, filter=self.filter)
+
+            # Convert the generator of RecordBatch objects to a list
+            batches_list = list(batches)
+
+            # Convert the list of RecordBatch objects to a Table
+            table = pa.Table.from_batches(batches_list)
+
+            return table
```

Comment:
Agreed to avoid Lance -> Pandas -> PyArrow trip.

@wjones127 - As the fragment can be pretty big, should we use `to_batches` instead of `to_table` here? https://lancedb.github.io/lance/read_and_write.html#iterative-read .

Reply:
@wjones127 and @c21 I looked at `to_batches` here - but that returns `record_batches`, and I couldn't identify how to get that into a PyArrow table. I think `to_table` is appropriate here, but I am curious about how to deal with large fragments

Reply:
@brent-anyscale The idea of using `to_batches()` is to stream data. It returns an `Iterable[RecordBatch]`, but if you want an `Iterable[Table]` you could simply do something like:

```python
def wrap_table(batch_iter):
    for batch in batch_iter:
        yield pa.Table.from_batches([batch])
```

Reply:
Converted to be similar to the Parquet datasource. Thanks @c21 for the pointers!

Reply:
Updated!